invokeai/app/services/batch_manager_storage.py

```diff
@@ -0,0 +1,514 @@
+from abc import ABC, abstractmethod
+from typing import cast
+import uuid
+import sqlite3
+import threading
+from typing import (
+    Any,
+    List,
+    Literal,
+    Optional,
+    Union,
+)
+import json
+
+from invokeai.app.invocations.baseinvocation import (
+    BaseInvocation,
+)
+from invokeai.app.services.graph import Graph
+from invokeai.app.models.image import ImageField
+
+from pydantic import BaseModel, Field, Extra, parse_raw_as
+
+invocations = BaseInvocation.get_invocations()
+InvocationsUnion = Union[invocations]  # type: ignore
+
+BatchDataType = Union[str, int, float, ImageField]
+
+
+class Batch(BaseModel):
+    data: list[dict[str, BatchDataType]] = Field(description="Mapping of node field to data value")
+    node_id: str = Field(description="ID of the node to batch")
```

Comment:
I know we had some discussions around this but I'm not sold on this shape.

Besides it feeling inefficient to keep repeating the field name, I can't imagine how to do zipped batches where the batch data is in multiple nodes.

I think this structure works for all cases:

```py
import itertools
import json
from typing import Union
from pydantic import BaseModel, validator, StrictStr, StrictInt, StrictFloat

class BatchData(BaseModel):
    node_id: str
    field_name: str
    items: list[Union[StrictStr, StrictInt, StrictFloat]]


class Batch(BaseModel):
    data: list[list[BatchData]]

    @validator("data")
    def validate_len(cls, v: list[list[BatchData]]):
        for batch_data in v:
            if any(len(batch_data[0].items) != len(i.items) for i in batch_data):
                raise ValueError("Zipped batch items must have all have same length")
        return v

    @validator("data")
    def validate_types(cls, v: list[list[BatchData]]):
        for batch_data in v:
            for datum in batch_data:
                for item in datum.items:
                    if not all(isinstance(item, type(i)) for i in datum.items):
                        raise TypeError("All items in a batch must have have same type")
        return v

    @validator("data")
    def validate_unique_field_mappings(cls, v: list[list[BatchData]]):
        paths: set[tuple[str, str]] = set()
        count: int = 0
        for batch_data in v:
            for datum in batch_data:
                paths.add((datum.node_id, datum.field_name))
                count += 1
        if len(paths) != count:
            raise ValueError("Each batch data must have unique node_id and field_name")
        return v


raw = json.dumps(
    {
        "data": [
            [
                {"node_id": "noise", "field_name": "seed", "items": [0, 1, 2, 3, 4]},
                {"node_id": "noise", "field_name": "width", "items": [512, 520, 528, 536, 544]},
            ],
            [{"node_id": "positive_conditioning", "field_name": "prompt", "items": ["cat", "dog", "lizard"]}],
        ]
    }
)

Batch.parse_raw(raw)

GoodBatch = Batch(data=[
    [
        BatchData(node_id="noise", field_name="seed", items=[0, 1, 2, 3, 4]),
        BatchData(node_id="noise", field_name="width", items=[512, 520, 528, 536, 544]),
    ],
    [
      BatchData(node_id="positive_conditioning", field_name="prompt", items=["cat", "dog", "lizard"])
    ]
])

UnequalSizes = Batch(data=[
    [
        BatchData(node_id="noise", field_name="seed", items=[0, 1, 2, 3, 4, 5]),
        BatchData(node_id="noise", field_name="width", items=[512, 520, 528, 536, 544]),
    ],
    [
      BatchData(node_id="positive_conditioning", field_name="prompt", items=["cat", "dog", "lizard"])
    ]
])

MismatchedTypes = Batch(data=[
    [
        BatchData(node_id="noise", field_name="seed", items=[0, 1, 2, 3, 4]),
        BatchData(node_id="noise", field_name="width", items=[512, 520, 528, 536, 544]),
    ],
    [
      BatchData(node_id="positive_conditioning", field_name="prompt", items=["cat", "dog", 2, "lizard"])
    ]
])
```

I *think* that's fully validated as well.

You'd need to be a bit more fancy than `itertools.product()`, but it covers all the use-cases.

Reply:
`itertools.product()` would still be plenty for this. I'll change it to this and just see how it feels